### PR TITLE
[cpu][vec512] improve bf16/fp16 load/store with mask for inductor

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116964
* #116963
* #116962
* __->__ #116961

Improve perf of vec512 bfloat16 (and also float16) load and store with partial vector lanes using masked load/store instead of via `memcpy` with aux buffer. In inductor CPU backend, we do load/store half (16) vector lanes for bfloat16 and float16.

Using the following micro-benchmark script for `layernorm + add`:
```python
import torch
import torch.nn as nn
from benchmark_helper import time_with_torch_timer

class AddLayernorm(nn.Module):
    def __init__(self, hidden_size):
        super().__init__()
        self.ln = nn.LayerNorm(hidden_size)

    def forward(self, hidden_states):
        return hidden_states + self.ln(hidden_states)

hidden_states = torch.randn(1, 512, 1024).to(torch.bfloat16)

with torch.no_grad():
    compiled_add_ln = torch.compile(add_ln)
    print(time_with_torch_timer(compiled_add_ln, hidden_states, iters=10000))
```

Measured on single-core `Intel(R) Xeon(R) Platinum 8358 CPU`.
Before: 1.39 ms
After: 498.66 us

cc @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by sanchitintel — APPROVED** [review]
LGTM, thanks!

This reminds me that I may have ignored basic vectorization principles elsewhere, and should revisit some code!

**Review by leslie-fang-intel — APPROVED** [review]
